saged/models.py

```diff
@@ -0,0 +1,136 @@
+        loss_type: The name of the loss function being used
+        accuracy_metric_type: The name of the accuracy metric being used
+        """
+        self.name = model_name
+        self.loss_type = loss_type
+        self.accuracy_metric_type = accuracy_metric_type
+        self.progress_type = progress_type
+
+        # Progress stores the number of iterations of type progress_type so far
+        self.val_progress = []
+        self.train_progress = []
+        self.val_loss = []
+        self.train_loss = []
+        self.val_metric = []
+        self.train_metric = []
+        self.other = {}
+
+    def add_progress(self,
+                     progress: int,
+                     loss: float,
+                     accuracy_metric: float,
+                     is_val: bool
+                    ) -> None:
+        """
+        Update the ModelResults with loss and accuracy metric information
+
+        Arguments
+        ---------
+        progress: The step, epoch, etc. that this entry corresponds to
+        loss: The value of the loss function at this time
+        accuracy_metric: The value of the accuracy metric at this time
+        is_val: Whether the results should be stored as validation metrics or training metrics
+        """
+        if is_val:
+            self.val_progress.append(progress)
+            self.val_loss.append(loss)
+            self.val_metric.append(accuracy_metric)
+        else:
+            self.train_progress.append(progress)
+            self.train_loss.append(loss)
+            self.train_metric.append(accuracy_metric)
+
+    def add_other(self,
+                  metric_name: str,
+                  metric_val: Union[int, float],
+                  metric_progress: int
+                 ) -> None:
+        """
+        Add information about an additional metric to the model
+
+        Arguments
+        ---------
+        metric_name: The name of the metric being recorded
+        metric_val: The value of the metric being recorded
+        metric_progress: The step, epoch, etc. that this entry corresponds to
+        """
+
+        if metric_name in self.other:
+            self.other['metric_name']['vals'].append(metric_val)
+            self.other['metric_name']['progress'].append(metric_progress)
+
+        else:
+            self.other['metric_name'] = {'vals': [metric_val],
+                                         'progress': [metric_progress]
+                                        }
+
+
+class ExpressionModel(ABC):
+    """ A model API similar to the scikit-learn API that will specify the
+    base acceptable functions for models in this module's benchmarking code
+    """
+
+    def __init__(self) -> None:
+        """
+        Standard model init function. We use pass instead of raising a NotImplementedError
+        here in case inheriting classes decide to call `super()`
+        """
+        pass
+
+    @abstractmethod
+    def fit(self, dataset: LabeledDataset) -> ModelResults:
```

Comment:
I think just having `fit` and `predict` methods here is good. For some models you'll want to implement additional methods like `decision_function`, `predict_proba`, etc, but not all models have those so keeping this class as general as possible seems like the way to go.

One thing you could consider is having `get_params` and `set_params` functions, which would allow saving (for instance) the weights of a trained NN or linear model and applying them to data later. But I'm not sure how this would work for something like k-nearest neighbors which doesn't really fit parameters, so maybe it's best to just add these in the inheriting classes.